### PR TITLE
The local coupling reformulation

### PR DESCRIPTION
**Pull request by pausz**
@i-Zaak is visiting us for a week and we discussed about removing the sparse matrix multiplication used to compute the local coupling term  in the models 'dfun'. 

The changes I just commited to this new branch mainly affect **where** the local coupling term  (local_coupling \* some_state_variable) is computed. 

@maedoc and @timpx:
I added a (dummy) variable in the models called lcvar to define local coupling variables. By default it is the first state variable of each model. It's completely wrong of course because:
 a) there could be more than one variable;
 b) there are some models that use a locally defiend auxiliary variable (difference of state variables, or a function of a state variable).

I did not update all the models, only where it was straigthforward and until the local coupling is completely functional.The local_coupling variable now represents the dot product of local_connectivity \* first_state_variable and computed directly in the simulator.

Some issues to discuss / consider:
- Consistent aming for the coupling variable (eg, global coupling variables --> gcvar and local coupling variables --> lcvar ?)
- Consistent naming for number of state variables, number of modes and  coupling variables. Make them model properties?

The local coupling might be best handled in a separate module because it needs to take care of the cases where there is more than one loca coupling variables; there are auxiliary variables; there are functions to be applied to the current state before computing the local coupling term.

### CODE REVIEW

**Comment by timpx** [comment]
Couldn't we directly reuse the functions of the existing coupling class for presynaptic computations?

File: tvb/simulator/models.py [hunk 1]
```diff
@@ -1080,13 +1088,13 @@ def dfun(self, state_variables, coupling, local_coupling=0.0):
...
                self.K12 * (numpy.dot(alpha, self.Bik) - xi) +
-               self.tau * (self.IE_i + c_0 + local_coupling * xi))
+               self.tau * (self.IE_i + c_0 + local_coupling))
 
         deta = (xi - self.b * eta + self.m_i) / self.tau
 
         dalpha = (self.tau * (alpha - self.f_i * alpha ** 3 / 3.0 - beta) +
                   self.K21 * (numpy.dot(xi, self.Cik) - alpha) +
-                  self.tau * (self.II_i + c_0 + local_coupling * xi))
+                  self.tau * (self.II_i + c_0 + local_coupling))
```
Comment (maedoc) [inline tvb/simulator/models.py:1103]: lol, expensive dot product computed twice. we're already faster now..

**Comment by maedoc** [comment]
good work! have you tried to run it?

**Comment by i-Zaak** [comment]
@timpx not really -- the weights for local coupling are sparse and broadcasting is not implemented for these :/ Currently trying to work around that.

**Comment by maedoc** [comment]
There are several places where we have `local_coupling=np.zeros(3)` followed by `local_coupling[0, :]`.

I'm not too happy about this implementation, 
- it can't be merged as is; given the age of the PR, 
- there's a nested for loop 
- the use of indices is prone to error, e.g. if you say `lcvar = [1]`, `local_coupling[0]` is not the first state variable.
- dense vs. sparse is orthogonal to local vs. global
- the test doesn't test much

As Paula summarized well, [`# This is flat out wrong. We need write down properly both global and local coupling.`](https://github.com/the-virtual-brain/tvb-library/pull/35/files#diff-fb1b78d850186af58a1b48b9d35aee05R100).

I'd like to work up an actual API for coupling that works nicely for both dense and sparse coupling structures, and allows them to be cleanly separated from the models.

**Comment by i-Zaak** [comment]
The parts with  `local_coupling=np.zeros(3)` can be safely removed and replaced by proper setting in simulator loop (based on number of coupling variables, and number of modes). Or refactored to "dummy surface coupling" class, which will be default for regional simulations.

The test doesn't test much: yes, please extend it:)

dense vs sparse -- should I rename it? Could make more sense.

the use of indices is prone to error -- yes, but can be tested against. I'd rather avoid more complex data structures for the state and coupling vectors.

The nested for loop is around expensive sparse matrix - vector multiplication, what is the problem with that?

**Comment by maedoc** [comment]
This isn't an urgent change, and this PR is now way out of date. We can retry this later on once API has stablilized.

**Pull request closed without merging**